File: frontend/web/admin-dashboard/app/lib/articleService.ts
const API_BASE_URL = "http://localhost:8000/api/articles";

// Types

export type ContentImage = {
  id: string;
  dataUrl: string;
  name: string;
};

export type Article = {
  id: string;
  title: string;
  genre: string;
  author: string;
  thumbnail: string;
  content: string;
  contentImages: ContentImage[];
  createdAt?: string;
};

export type CreateArticlePayload = {
  title: string;
  genre: string;
  author: string;
  thumbnail: string;
  content: string;
  contentImages: ContentImage[];
};

export type UpdateArticlePayload = CreateArticlePayload;

// Helper: Auth Header

async function getAuthHeaders(): Promise<HeadersInit> {
  const user = (await import("firebase/auth")).getAuth().currentUser;

  if (!user) throw new Error("User not authenticated");

  const token = await user.getIdToken();

  return {
    "Content-Type": "application/json",
    Authorization: `Bearer ${token}`,
  };
}

// API Calls

// Get all articles
export async function fetchArticles(): Promise<Article[]> {
  const headers = await getAuthHeaders();

  const res = await fetch(API_BASE_URL, {
    method: "GET",
    headers,
  });

  if (!res.ok) {
    throw new Error("Failed to fetch articles");
  }

  return res.json();
}

// Create article
export async function createArticle(
  data: CreateArticlePayload
): Promise<Article> {
  const headers = await getAuthHeaders();

  const res = await fetch(API_BASE_URL, {
    method: "POST",
    headers,
    body: JSON.stringify(data),
  });

  if (!res.ok) {
    const error = await res.json();
    throw new Error(error.detail || "Failed to create article");
  }

  return res.json();
}

// Get single article
export async function getArticle(id: string): Promise<Article> {
  const headers = await getAuthHeaders();

  const res = await fetch(`${API_BASE_URL}/${id}`, {
    method: "GET",
    headers,
  });

  if (!res.ok) {
    throw new Error("Article not found");
  }

  return res.json();
}

// Update article
export async function updateArticle(
  id: string,
  data: UpdateArticlePayload
): Promise<Article> {
  const headers = await getAuthHeaders();

  const res = await fetch(`${API_BASE_URL}/${id}`, {
    method: "PATCH",
    headers,
    body: JSON.stringify(data),
  });

  if (!res.ok) {
    const error = await res.json();
    throw new Error(error.detail || "Failed to update article");
  }

  return res.json();
}

// Delete article
export async function deleteArticle(id: string): Promise<void> {
  const headers = await getAuthHeaders();

  const res = await fetch(`${API_BASE_URL}/${id}`, {
    method: "DELETE",
    headers,
  });

  if (!res.ok) {
    throw new Error("Failed to delete article");
  }
}

File: frontend/mobile/lib/Article/article_service.dart
import 'dart:convert';
import '../core/network/backend_connector.dart';

class ContentImage {
  final String id;
  final String dataUrl;
  final String name;

  ContentImage({required this.id, required this.dataUrl, required this.name});

  factory ContentImage.fromJson(Map<String, dynamic> json) => ContentImage(
        id: json['id'] as String? ?? '',
        dataUrl: json['dataUrl'] as String? ?? '',
        name: json['name'] as String? ?? '',
      );
}

class ArticleModel {
  final String id;
  final String title;
  final String genre;
  final String author;
  final String thumbnail;
  final String content;
  final List<ContentImage> contentImages;

  ArticleModel({
    required this.id,
    required this.title,
    required this.genre,
    required this.author,
    required this.thumbnail,
    required this.content,
    required this.contentImages,
  });

  factory ArticleModel.fromJson(Map<String, dynamic> json) => ArticleModel(
        id: json['id'] as String? ?? '',
        title: json['title'] as String? ?? '',
        genre: json['genre'] as String? ?? '',
        author: json['author'] as String? ?? '',
        thumbnail: json['thumbnail'] as String? ?? '',
        content: json['content'] as String? ?? '',
        contentImages: (json['contentImages'] as List<dynamic>? ?? [])
            .map((e) => ContentImage.fromJson(Map<String, dynamic>.from(e as Map)))
            .toList(),
      );
}

class ArticleService {
  static final _backend = BackendConnector.instance;

  static Future<List<ArticleModel>> fetchPublished() async {
    final res = await _backend.get('/articles');
    if (res.statusCode != 200) throw Exception('Failed to load articles');
    final data = jsonDecode(res.body) as Map<String, dynamic>;
    final list = data['articles'] as List<dynamic>;
    return list.map((e) => ArticleModel.fromJson(Map<String, dynamic>.from(e as Map))).toList();
  }
}
